### PR TITLE
CTE Refactor for robustness

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
This bug showed problems with handling lookups and suggested columns
https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=45896

#### Related Pull Requests
* 

#### Changes
Don't use QuerySelect.getTableInfo() to turn CTE into table info.  Use "normal" mechanism of leaving query data structures and calling getSQL() on the whole thing at once.
Introduced/repurposed an intermediate wrapper QuerySelect.QueryWithWrapper to isolate separate usages of a CTE.

### CODE REVIEW

File: query/src/org/labkey/query/QueryModule.java [hunk 1]
```diff
@@ -341,7 +341,7 @@ public Set<Class> getIntegrationTests()
         return Set.of(
             ModuleReportCache.TestCase.class,
             OlapController.TestCase.class,
-            Query.QueryTestCase.class,
+            //Query.QueryTestCase.class,
```
Comment (labkey-jeckels) [inline query/src/org/labkey/query/QueryModule.java]: Can this be completely deleted with the new .jsp test?

File: query/src/org/labkey/query/QueryTestCase.jsp [hunk 2]
```diff
@@ -0,0 +1,1763 @@
...
+    {
+        private static final String[] COLUMNS = new String[] {"d", "seven", "twelve", "day", "month", "date", "duration", "guid"};
+        private static final JdbcType[] TYPES = new JdbcType[] {JdbcType.DOUBLE, JdbcType.INTEGER, JdbcType.INTEGER, JdbcType.VARCHAR, JdbcType.VARCHAR, JdbcType.TIMESTAMP, JdbcType.VARCHAR, JdbcType.GUID};
+        private static final String[] days = new String[] {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};
+        private static final String[] months = new String[] {"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"};
+
+        private final String[][] data;
+        private final RowMapFactory<Object> _rowMapFactory;
+
+        // UNDONE: need some NULLS in here
```
Comment (labkey-jeckels) [inline query/src/org/labkey/query/QueryTestCase.jsp:99]: Is this for now, or the hypothetical future?

File: query/src/org/labkey/query/sql/Query.java [hunk 3]
```diff
@@ -2646,6 +2526,173 @@ public void testContainerAnnotation() throws Exception
         }
 
 
+        static SqlTest[] cteTests = new SqlTest[]
```
Comment (labkey-jeckels) [inline query/src/org/labkey/query/sql/Query.java]: Can this be deleted? Has it been moved to the JSP?

**Review by labkey-jeckels — APPROVED** [review]
This is tough to review from a diff perspective because a lot of important the code (including test coverage) moved between files. Let me know if there are specific areas that I should focus on.

File: query/src/org/labkey/query/sql/Query.java (hunk 3 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/Query.java]: Thanks, I thought that was deleted!

File: query/src/org/labkey/query/QueryTestCase.jsp (hunk 2 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/QueryTestCase.jsp:99]: old comment

**Comment by labkey-matthewb** [comment]
> 

Yes, sorry about that.  I'm happy to walk through it.  Always nice to have more than one person know some of the Query code.

**Comment by labkey-matthewb** [comment]
> > 
> 
> Yes, sorry about that. I'm happy to walk through it. Always nice to have more than one person know some of the Query code.

Mostly testCTE() is new in the QueryTestCase.jsp and then the CTE code looks a lot the same, but is actually pretty different!

**Pull request merged**